### PR TITLE
nixos/lightdm-gtk-greeter: make it overridable by other greeters

### PR DESCRIPTION
**Pull request by mnacamura**
###### Motivation for this change

So that we do not have to care about the order to import greeters modules: https://github.com/NixOS/nixpkgs/blob/bb3f7d14a17e88e4341872c55d8b32127c6f8101/nixos/modules/services/x11/display-managers/lightdm.nix#L70-L76

<s>Moreover, I added the second commit in which gtk greeter is disabled automatically if any other greeter is enabled.</s> Currently, every time when adding a greeter module, one has to add
```
services.xserver.displayManager.lightdm.greeters.gtk.enable = false;
```
in the config section. By this PR, no need to do it.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by mnacamura** [comment]
Is [this build error on the second commit](https://gist.github.com/GrahamcOfBorg/32bfe984071509a3de67e9129c4ce294) relevant to this PR? Do I need to fix something?

**Comment by Infinisil** [comment]
This is definitely not the way to implement this.

I think this can be done by simply removing `gtk.enable = false;` from `mini.nix`, and replacing the `mkDefault` at the bottom of `gtk.nix` with `mkOptionDefault`. Then any greeter will be able to override the gtk default, and only one can be enabled because the options `services.xserver.displayManager.lightdm.greeter.{package,name}` can only be assigned once.

**Comment by mnacamura** [comment]
@Infinisil Rewrote it based on your idea. Thanks.

File: nixos/modules/services/x11/display-managers/lightdm-greeters/gtk.nix [hunk 1]
```diff
@@ -148,14 +148,19 @@ in
...
-      name = "lightdm-gtk-greeter";
-    };
+      services.xserver.displayManager.lightdm.greeter = mkOptionDefault {
+        package = wrappedGtkGreeter;
+        name = "lightdm-gtk-greeter";
+      };
 
-    environment.etc."lightdm/lightdm-gtk-greeter.conf".source = gtkGreeterConf;
+    })
+    (mkIf (ldmcfg.greeter.name == "lightdm-gtk-greeter") {
```
[the remark this answers is not in the document]
Comment (mnacamura) [reply nixos/modules/services/x11/display-managers/lightdm-greeters/gtk.nix]: Fixed.

**Comment by mmahut** [comment]
Are there any updates on this pull request, please?

**Comment by lucasew** [comment]
Is it still relevant?

**Pull request closed without merging**